### PR TITLE
Remove EnsureCommunicatingAfterTransitions

### LINKED ISSUE #6497: Alternatives for current ensure_communicating
We've recently been refactoring the worker state machine (https://github.com/dask/distributed/issues/5736)

The new system works great for single-task-based transitions but feels awkward as soon as we're trying to treat groups or batches of tasks similarly which is particularly important for fetching remote data. We are trying to not perform a network request per key but rather try to perform a request per worker and try to get as many keys of it as reasonable.

This mechanism worked so far by running `Worker.(_)ensure_communicating` frequently which iterates over all tasks and schedules a fetch/gather_dep (pseudo code below)

```python
def ensure_communicating(...):
    skip = []
    while self.data_needed and free_slots_for_fetch:
        ts = self.data_needed.pop() # highest priority task
        worker = pick_good_worker(ts) # not busy, maybe localhost, etc.
        if not worker:
            skip.add(ts)
        more_keys_on_worker = select_keys_for_gather(worker)
        schedule_gather_dep(ts, more_keys_on_worker, worker) # This schedules a coroutine and actually fetches the data
    self.data_needed.update(skip) # data needed is a heap so this is non-linear
```

This has a couple of problems
1. It's unclear when to run this function. Past iterations just ran this frequently (on every incoming request and in various code places where we guess there may be something that needs transitioning
2. We're iterating over all keys, all the time. This is bad for performance although not critical on the worker https://github.com/dask/distributed/pull/6462#issuecomment-1138764541
3. If we're calling this too soon, e.g. within a released->fetch transition, we may get poor scheduling decisions by fetching tasks prematurely since we didn't work off all fetch recommendations before calling this. For this reason, the `EnsureCommunicatingAfterTransitions ` instruction was introduced (see also https://github.com/dask/distributed/pull/6462) to mimic earlier behavior

## Proposal

I think we can get rid of `Worker.data_needed` entirely. Since we're storing everything in heaps, by now, `Worker.data_needed_per_worker` carries everything we need if we change how ensure_communicating works which might have other interesting side effects.

To illustrate this, consider the following two helper functions

```python
def get_most_important_worker_to_fetch_from(self):
    """
    Get the remote worker that holds the _most important_ data we should fetch. This is typically inferred by looking at task priorities.
    This could also be used to modify what we consider most important worker to fetch from, e.g. due to load/business, memory pressure, etc.
    """

    # Linear in num_workers
    res = None
    prio = None
    for w, heap in self.data_needed_per_worker.items():
        nprio = heap.peek().priority
        if w not in self.busy_workers and (res is None or nprio > prio):
            res = w
            prio = nprio
    return res

def get_tasks_for_worker(self, w, limit=42):
    """Get a list of tasks on worker `w` such that combined they have at most `limit` estimated bytes"""
    # This is more or less the current `Worker._select_keys_for_gather`
    # Linear in number of tasks in local state fetch on remote worker
    res = []
    size = 0
    for ts in self.data_needed_per_worker[w]:
        size += ts.get_nbytes()
        res.append(ts)
        if size > limit:
            break
    return res
```

With these two helpers we should be able to construct a new, simplified version if we switch our loop from tasks to workers

```python
def _new_ensure_communicate_with_tasks_by_worker():
    if network_limits_not_reached and worker := self.get_most_important_worker_to_fetch_from():
        tasks_to_fetch = self.get_tasks_for_worker(worker)
        instruction = GatherDep(
            worker,
            tasks_to_fetch,
        )
```

or even change the `GatherDep` instruction to worker-only and delay the decision of what tasks to fetch until this instruction is acted on, which typically only happens after all transitions in `Worker.transitions` are worked off

```python
def _new_ensure_communicate_with_tasks_by_worker():
    if network_limits_not_reached and worker := self.get_most_important_worker_to_fetch_from():
        instruction = GatherDep(worker)

def gather_dep(self, worker):
    tasks_to_fetch = self.get_tasks_for_worker(worker)
    ...
```

cc @crusaderky

### PR DESCRIPTION
**Pull request by crusaderky**
Partially closes #6497

#6165 introduced this hack, for the sake of being functionally identical to the previous code.
This is a cleaner redesign which is conceptually the same.

### CODE REVIEW

**Comment by crusaderky** [comment]
The transition log has changed from
```
           - ('x', 'ensure-task-exists', 'released')
           - ('x', 'released', 'fetch', 'fetch', {})
           - ('gather-dependencies', 'tcp://127.0.0.1:53985', {'x'})
           - ('x', 'fetch', 'flight', 'flight', {})
```
to
```
           - ('x', 'ensure-task-exists', 'released'),
           - ('gather-dependencies', 'tcp://127.0.0.1:53985', {'x'}),
           - ('x', 'released', 'fetch', 'fetch', {'x': ('flight', 'tcp://127.0.0.1:53985')}),
           - ('x', 'fetch', 'flight', 'flight', {}),
```
This... is correct, but it's counter-intuitive as hell.
What's happening is that
1. `transition_released_fetch` starts. It sets `ts.state = "fetch"`, adds it to `data_needed`, and internally invokes `_ensure_communicating`.
2. `_ensure_communicating` removes ts from `data_needed`, prints its own log line `gather-dependencies`, and returns a recommendation to transition to flight, together with a `GatherDep` instruction
3. `transition_released_fetch` returns
4. `_transition` logs the released->fetch transition and returns the recommendations, instructions generated by `_ensure_communicating`
5. `_transitions` calls `_transition` again for fetch->flight and logs the outcome.

Again, all this is *correct*, but it's extremely confusing. It took me an unhealthy amount of time to figure out why the `gather-dependencies` log line appeared before `('x', 'released', 'fetch', 'fetch')`. And I've been trawling into the state machine for many weeks now.

Not sure if we can/want to do anything about this?

**Comment by crusaderky** [comment]
3 tests currently failing:

#6303:
```
All 7 runs failed: test_basic (distributed.dashboard.tests.test_worker_bokeh)
```


This is worrying and needs investigating thoroughly:
```
2 out of 8 runs failed: test_submit_after_failed_worker_async (distributed.tests.test_failed_workers)
    def validate_task_fetch(self, ts):
>       assert ts in self.data_needed
E       AssertionError
```

This looks to be timings-sensitive; needs redesigning:
```
3 out of 8 runs failed: test_decide_worker_coschedule_order_neighbors[nthreads1-0] (distributed.tests.test_scheduler)
3 out of 8 runs failed: test_decide_worker_coschedule_order_neighbors[nthreads1-1] (distributed.tests.test_scheduler)
1 out of 8 runs failed: test_decide_worker_coschedule_order_neighbors[nthreads1-4] (distributed.tests.test_scheduler)
```

**Comment by fjetter** [comment]
> This... is correct, but it's very counter-intuitive.
> What's happening is that

As I'm arguing in https://github.com/dask/distributed/pull/6442#issuecomment-1143860614 I believe this log should simply be removed

**Comment by fjetter** [comment]
I could reproduce the issue [about the assumptions in the test about which keys to be fetched](https://github.com/dask/distributed/pull/6472#pullrequestreview-994648986). This is not entirely obvious from the tests and I consider this a bit concerning.

It's also not about any priorities, ordering, etc. but rather that we're using an unordered set for `ts.dependencies` which then triggers transitions here

https://github.com/dask/distributed/blob/6d85a8536ee02fd2a9ae52f50de030d1cc7afe7e/distributed/worker.py#L2125-L2129

This randomness was previously buffered by the delayed _ensure_communicating. I don't feel great about introducing randomness to our scheduling (e.g. there is not even a key-based tie breaker). This entire refactoring effort is supposed to help us make things more deterministic.

**Comment by fjetter** [comment]
I opened https://github.com/dask/distributed/issues/6497 with a suggestion on how to move forward with ensure_communicating. Maybe it's worth putting this PR on ice until #6497 is settled

**Comment by crusaderky** [comment]
I'm against implementing #6497 before #6476, for reasons that should be obvious.
I can put this on ice and go straight for #6476; the only caveat is that there will be an ugly private call from `BaseWorker._handle_instructions` to `BaseWorker.state._transitions`. I think it's tolerable as long as it's not there to last.

**Comment by crusaderky** [comment]
his has been parked until a new design is agreed upon in #6497.

**Comment by crusaderky** [comment]
The PR has been rewritten from scratch and is now ready for review and merge.
It exacerbates the O(n^2*logn) condition described above; this is fixed in #6587.

File: distributed/worker_state_machine.py [hunk 1]
```diff
@@ -2547,10 +2547,7 @@ def _handle_gather_dep_success(self, ev: GatherDepSuccessEvent) -> RecsInstrs:
                 self.has_what[ev.worker].discard(ts.key)
                 recommendations[ts] = "fetch"
 
-        return merge_recs_instructions(
-            (recommendations, []),
-            self._ensure_communicating(stimulus_id=ev.stimulus_id),
-        )
```
Comment (fjetter) [inline distributed/worker_state_machine.py:2553]: I love that this is not everywhere anymore ❤️

**Review by fjetter — APPROVED** [review]
This is great!

**Pull request merged**